lnst/RecipeCommon/Perf/Measurements/LinuxPerfMeasurement.py

```diff
@@ -0,0 +1,240 @@
+
+    @property
+    def end_timestamp(self):
+        return self._end_timestamp
+
+    def time_slice(self, start, end):
+        return self
+
+
+class AggregatedLinuxPerfMeasurementResults(BaseMeasurementResults):
+
+    _individual_results: list[LinuxPerfMeasurementResults] = []
+
+    def __init__(self, result: Optional[LinuxPerfMeasurementResults] = None):
+        if result:
+            self._individual_results = [result]
+
+    @property
+    def individual_results(self) -> list[LinuxPerfMeasurementResults]:
+        return self._individual_results
+
+    def add_results(self, result: LinuxPerfMeasurementResults):
+        self._individual_results.append(result)
+
+
+class LinuxPerfMeasurement(BaseMeasurement):
+    _MEASUREMENT_VERSION: int = 1
+
+    hosts: list[Host]
+    intr_filename: str
+    intr_cpus: list[int]
+    iperf_filename: str
+    iperf_cpus: list[int]
+    _start_timestamp: float
+    _end_timestamp: float
+    _data_folder: str
+
+    _version: Optional[dict[str, Any]] = None
+    _collection_index: int = 0
+    _jobs: list[Job] = []
+
+    def __init__(
+        self,
+        hosts: list[Host],
+        intr_filename: str,
+        intr_cpus: list[int],
+        iperf_filename: str,
+        iperf_cpus: list[int],
+        data_folder: str,
+        recipe_conf: Any = None,
+    ):
+        super().__init__(recipe_conf)
+
+        self.hosts = hosts
+        self.intr_filename = intr_filename
+        self.intr_cpus = intr_cpus
+        self.iperf_filename = iperf_filename
+        self.iperf_cpus = iperf_cpus
+
+        self._data_folder = data_folder
+
+        # create output folders
+        for host in self.hosts:
+            try:
+                os.mkdir(os.path.join(self._data_folder, host.hostid))
+            except FileExistsError:
+                pass
+
+    @property
+    def version(self) -> Optional[dict[str, Any]]:
+        if self._version:
+            return self._version
+
+        perf_version: Optional[str] = self._get_host_perf_version()
+        if perf_version is None:
+            return None
+
+        self._version = {
+            "measurement_version": self._MEASUREMENT_VERSION,
+            "host_perf_version": perf_version,
+        }
+        return self._version
+
+    def _get_host_perf_version(self) -> Optional[str]:
+        if not len(self.hosts):
+            raise Exception(
+                "No hosts in LinuxPerfMeasurement while getting perf version"
+            )
+
+        version_job = self.hosts[0].run("perf --version", job_level=ResultLevel.DEBUG)
+        if version_job.passed:
+            match = re.match(r"perf version (.+?)", version_job.stdout)
+            if match:
+                return match.group(1)
+        return None
+
+    def _generate_configurations(self):
+        for host in self.hosts:
+            yield (host, self.intr_filename, self.intr_cpus)
+            yield (host, self.iperf_filename, self.iperf_cpus)
```

Comment:
Since this is supposed to be used as a "common" code for recipes not necessarilly coming from ENRT recipe set, I think the concept of dev_intr_cpu and perf_tool_cpu should be unified and instead have a list of cpus to be monitored. With that we will not need to include 'intr' or 'perf' substrings in the generated files, only the CPU id. This can be done as a followup pull request, just adding a note here.

Reply:
I thought it mattered on what CPU the data was captured? What would someone had to do in order to distinguish between the output files, if they were named after the CPU they ran on?

But I agree, I could rework the `LinuxPerfMeasurement` class to accept a list of CPUs.

Reply:
Of course, we want to distinguish the perf.data for individual cpus. The thing is that if an lnst user creates a recipe that DOES NOT use the `BaseEnrtRecipe` and only the `RecipeCommon` (which is independent of ENRT), there are no `dev_intr_cpu` and `perf_tool_cpu` parameters. So, to abstract this, we can substitute this with e.g. cpu0 and cpu5. The ENRT developer knows that cpu0 was used as dev_intr_cpu and cpu5 was used as perf_tool_cpu. The non-ENRT developer can use this differently.

So, you would have filenames:
```
cpu0.data.1.2022-04-23_14:28:32
cpu5.data.1.2022-04-23_14:28:32
```
instead of:
```
intr.data.1.2022-04-23_14:28:32
iperf.data.1.2022-04-23_14:28:32
```

Reply:
Hmm, I looked at the code again and actually your argument is kind of valid. Still I think we should do the abstraction and remove the dev_intr_cpu and perf_tool_cpu relation from the code.

To elaborate more on my idea:
* the `linuxperf_cpus`
  * will substitute the `LinuxPerfMeasurement.intr_cpus` and `LinuxPerfMeasurement.iperf_cpus`
  * will be a list of lists of cpus, e.g.
    * `linuxperf_cpus = [ [0, 5] ]` ... would run one perf-record for both cpu0 and cpu5
    * `linuxperf_cpus = [ [0], [5] ]` ... would run two perf records individually for cpu0 and cpu5
    * `linuxperf_cpus = [ [0,1,2,3] [4,5,6,7] ]` ... would run two perf records individually for the lists of cpus

The filenames will be generated based on linuxperf_cpus items:
* `linuxperf_cpus = [ [0], [5] ]`
  * cpus.0.data...
  * cpus.5.data...
* `linuxperf_cpus = [ [0,1,2,3] [4,5,6,7] ]`
  * cpus.0.1.2.3.data...
  * cpus.4.5.6.7.data...

So the `generate_configurations` code would look like this:
```
    for host in self.hosts:
        for cpu_list in self.linuxperf_cpus:
            yield (host, cpu_list)
```

@olichtne does this make sense?

Reply:
yes, i like this, additionally the ENRT recipe integration of this could then be:

```
@property
def linuxperf_cpus(self):
    return [self.dev_intr_cpus, self.perf_tool_cpus]
```

Reply:
@olichtne in which file should I put this property?

Reply:
You should also update the lnst/Recipes/ENRT/MeasurementGenerators/LinuxPerfMeasurementGenerator.py where there are still filename parameters:
```python
class LinuxPerfMeasurementGenerator(BaseMeasurementGenerator):                  
    do_linuxperf_measurement = BoolParam(default=False)                         
    linuxperf_intr_fname = StrParam(default="intr.data")                        
    linuxperf_iperf_fname = StrParam(default="iperf.data")                      
```

And once you add the new property, replace the following code:
```python
        for combination in combinations:                                        
            measurement: BaseMeasurement = LinuxPerfMeasurement(                
                self.matched,                                                   
                [self.params.dev_intr_cpu, self.params.perf_tool_cpu],          
                data_folder=linuxperf_data_folder,                              
                recipe_conf=config,                                             
            )                                                                   
```

Reply:
@olichtne you meant this, right?
```py
@property
def linuxperf_cpus(self):
    return [self.params.dev_intr_cpu, self.params.perf_tool_cpu]
```

Reply:
@jtluka I suppose I could use the newly created `linuxperf_cpus` property instead of the `[self.params.dev_intr_cpu, self.params.perf_tool_cpu]`, right?

Reply:
> @olichtne you meant this, right?
> 
> ```python
> @property
> def linuxperf_cpus(self):
>     return [self.params.dev_intr_cpu, self.params.perf_tool_cpu]
> ```

yes

Reply:
resolved in 49a0b807